File: 215.kth-largest-element-in-an-array.ts
/*
 * @lc app=leetcode id=215 lang=typescript
 *
 * [215] Kth Largest Element in an Array
 */

// @lc code=start
function findKthLargest(nums: number[], k: number): number {
  k = nums.length - k;
  function quickSortToK(start: number, end: number): number {
    if (end - start <= 0) return nums[k];
    let pivot = nums[end];
    let left = start;
    for (let i = start; i < end; i++) {
      if (nums[i] < pivot) {
        let temp = nums[i];
        nums[i] = nums[left];
        nums[left] = temp;
        left++;
      }
    }
    nums[end] = nums[left];
    nums[left] = pivot;
    if (k < left) {
      quickSortToK(start, left - 1);
    }
    if (k > left) {
      quickSortToK(left + 1, end);
    }
    return nums[k];
  }
  return quickSortToK(0, nums.length - 1);
}
// @lc code=end

File: 215.kth-largest-element-in-an-array.py
#
# @lc app=leetcode id=215 lang=python3
#
# [215] Kth Largest Element in an Array
#

# @lc code=start
from typing import List


class Solution:
    def partition(self, nums, s, e):
        key = nums[e]
        j = s
        for i in range(s, e):
            if nums[i] <= key:
                nums[j], nums[i] = nums[i], nums[j]
                j += 1
        nums[j], nums[e] = nums[e], nums[j]
        return j

    def findKthLargest(self, nums: List[int], k: int) -> int:
        kthIndex = len(nums) - k
        s = 0
        e = len(nums)-1
        while s < e:
            pivot = self.partition(nums, s, e)
            if pivot < kthIndex:
                s = pivot+1
            elif pivot > kthIndex:
                e = pivot-1
            else:
                break
        return nums[kthIndex]
# @lc code=end
